src/transformers/models/maskformer/modeling_maskformer.py

```diff
@@ -0,0 +1,1638 @@
+        )
+        # TODO return a tuple is not a good idea, torch.summary will fail
+        return detr_output.hidden_states
+
+
+class MaskFormerSegmentationModule(nn.Module):
+    def __init__(self, config: MaskFormerConfig):
+        super().__init__()
+        # + 1 because we add the "null" class
+        self.mask_classification = config.mask_classification
+        self.class_predictor = nn.Linear(config.hidden_size, config.num_classes + 1)
+        self.mask_embedder = MLP(config.hidden_size, config.hidden_size, config.mask_feature_size)
+
+    def forward(self, decoder_outputs: Tuple[Tensor], pixel_embbeddings: Tensor) -> Dict[str, Tensor]:
+        last_decoder_output: Tensor = decoder_outputs[-1]
+        mask_embeddings: Tensor = self.mask_embedder(last_decoder_output)
+        out: Dict[str, Tensor] = {}
+
+        if self.mask_classification:
+            classes: Tensor = self.class_predictor(last_decoder_output)
+            out.update({"pred_logits": classes})
+        # sum up over the channels
+        binary_masks: Tensor = torch.einsum("bqc,   bchw -> bqhw", mask_embeddings, pixel_embbeddings)
+        out.update({"pred_masks": binary_masks})
+        return out
+
+
+@dataclass
+class MaskFormerOutput(ModelOutput):
+    pred_logits: torch.FloatTensor
+    pred_masks: torch.FloatTensor = None
+    loss: Optional[torch.FloatTensor] = None
+    loss_dict: Optional[Dict] = None
+
+
+def upsample_like(x: Tensor, like: Tensor, mode: str = "bilinear") -> Tensor:
+    """An utility function that upsamples `x` to match the dimension of `like`
+
+    Args:
+        x (Tensor): The tensor we wish to upsample
+        like (Tensor): The tensor we wish to use as size target
+        mode (str, optional): The interpolation mode. Defaults to "bilinear".
+
+    Returns:
+        Tensor: The upsampled tensor
+    """
+    _, _, h, w = like.shape
+    upsampled: Tensor = F.interpolate(
+        x,
+        size=(h, w),
+        mode=mode,
+        align_corners=False,
+    )
+    return upsampled
+
+
+class MaskFormerModel(PreTrainedModel):
+    config_class = MaskFormerConfig
+    base_model_prefix = "model"
+    main_input_name = "pixel_values"
+
+    def __init__(self, config: MaskFormerConfig):
+        super().__init__(config)
+        self.pixel_level_module = MaskFormerPixelLevelModule(config)
+        self.transformer_module = MaskFormerTransformerModule(
+            in_features=self.pixel_level_module.backbone.get_outputs_shape()[-1], config=config
+        )
+        self.segmentation_module = MaskFormerSegmentationModule(config)
+        self.matcher = MaskFormerHungarianMatcher(
+            cost_class=1.0, cost_dice=config.dice_weight, cost_mask=config.mask_weight
+        )
+
+        losses = ["labels", "masks"]
+        # TODO I don't like too aggressive abbreviations in the naming
+        # what is 'ce'? It's cross entropy but it's not clear
+        # ask if we should follow a more exhaustive naming convention
```

Comment:
Yes definitely. It's the `classification_loss` for instance if it's cross entropy, which is way clearer.